src/ucp/tag/offload/rndv.c

```diff
@@ -0,0 +1,232 @@
+#include <ucp/rndv/proto_rndv.h>
+#include <ucp/rndv/proto_rndv.inl>
+#include <ucp/tag/offload.h>
+#include <ucp/tag/tag_rndv.h>
+
+
+/* Init HW rendezvous as a single protocol which does not require operation on
+ * remote side (which is the case when tag is matched - everyting is done by the
+ * HW/FW).
+ */
+static ucs_status_t
+ucp_tag_rndv_offload_proto_init(const ucp_proto_init_params_t *init_params)
+{
+    ucp_worker_h worker        = init_params->worker;
+    ucp_context_h context      = worker->context;
+    ucp_ep_config_t *ep_config = &ucs_array_elem(&worker->ep_config,
+                                                 init_params->ep_cfg_index);
+    ucp_proto_single_init_params_t params = {
+       .super.super         = *init_params,
+       .super.latency       = 0,
+       .super.overhead      = 40e-9,
+       .super.cfg_thresh    = context->config.ext.rndv_thresh,
+       .super.cfg_priority  = 60,
+       .super.min_length    = ucp_ep_tag_offload_min_rndv_thresh(ep_config),
+       .super.max_length    = ep_config->tag.offload.max_rndv_zcopy,
+       .super.min_iov       = 0,
+       .super.min_frag_offs = UCP_PROTO_COMMON_OFFSET_INVALID,
+       .super.max_frag_offs = UCP_PROTO_COMMON_OFFSET_INVALID,
+       .super.max_iov_offs  = ucs_offsetof(uct_iface_attr_t,
+                                           cap.tag.rndv.max_iov),
+       .super.hdr_size      = 0,
+       .super.send_op       = UCT_EP_OP_RNDV_ZCOPY,
+       .super.memtype_op    = UCT_EP_OP_LAST,
+       .super.flags         = UCP_PROTO_COMMON_INIT_FLAG_SEND_ZCOPY |
+                              UCP_PROTO_COMMON_INIT_FLAG_RECV_ZCOPY |
+                              UCP_PROTO_COMMON_INIT_FLAG_SINGLE_FRAG,
+       .super.exclude_map   = 0,
+       .lane_type           = UCP_LANE_TYPE_TAG,
+       .tl_cap_flags        = UCT_IFACE_FLAG_GET_ZCOPY |
+                              UCT_IFACE_FLAG_TAG_RNDV_ZCOPY
+    };
+
+    if (!ucp_tag_rndv_check_op_id(init_params) ||
+        (init_params->select_param->dt_class != UCP_DATATYPE_CONTIG)) {
+        return UCS_ERR_UNSUPPORTED;
+    }
+
+    return ucp_proto_single_init(&params);
+}
+
+static void
+ucp_tag_rndv_offload_completion(uct_completion_t *uct_comp)
+{
+    ucp_request_t *req = ucs_container_of(uct_comp, ucp_request_t,
+                                          send.state.uct_comp);
+
+    ucp_send_request_id_release(req);
+    ucp_proto_request_zcopy_complete(req, uct_comp->status);
+}
+
+static ucs_status_t
+ucp_tag_rndv_offload_request_init(ucp_request_t *req, ucp_md_map_t md_map,
+                                  uct_completion_callback_t comp_func,
+                                  unsigned uct_reg_flags, unsigned dt_mask)
+{
+    const ucp_proto_single_priv_t *spriv = req->send.proto_config->priv;
+    ucs_status_t status;
+
+    status = ucp_ep_resolve_remote_id(req->send.ep, spriv->super.lane);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    ucs_assert(md_map != 0);
```

Comment:
why needed?

Reply:
otherwise it means that rndv offload can't be used

Reply:
wouldn't it fail in ucp_proto_request_zcopy_init?

Reply:
i do not see a similar assert there, so it seems to be useful.
remove?

Reply:
actually this assert should fail:
https://github.com/openucx/ucx/blob/d4bdc86aa18abbacaea3fa2220ba6b8b1ed39896/src/ucp/dt/datatype_iter.inl#L474
from ucp_proto_zcopy_single_progress->ucp_datatype_iter_next_iov

Reply:
ic, though it does not seem really obvious and can be changed unrelatively to hwtm. So, i'd still keep it.
remove?

Reply:
yes. In theory there could be HWTM transport that does not need uct memh.

Reply:
ok